Sources/System/Sockets/SocketDescriptor+Messages.swift

```diff
@@ -0,0 +1,539 @@
+    precondition(i._offset >= 0 && i._offset + _headerSize <= _endOffset,
+                 "Invalid index")
+    guard let length = _length(at: i) else { return endIndex }
+    return Index(_offset: i._offset + length)
+  }
+
+  /// Returns the message at the given position, which must be a valid index
+  /// in this collection.
+  ///
+  /// The returned value merely refers to a slice of the entire buffer, so
+  /// it contains a shared regerence to it.
+  ///
+  /// To reduce memory use and to prevent unnecessary copy-on-write copying, do
+  /// not save `Message` values -- instead, copy out the data you need to hold
+  /// on to into standalone storage.
+  public subscript(position: Index) -> Message {
+    guard let _ = _length(at: position) else {
+      preconditionFailure("Invalid index")
+    }
+    return Element(_base: self, offset: position._offset)
+  }
+}
+
+// @available(macOS 9999, iOS 9999, watchOS 9999, tvOS 9999, *)
+extension SocketDescriptor.AncillaryMessageBuffer.Message {
+  internal var _header: CInterop.CMsgHdr {
+    _base._withUnsafeBytes { buffer in
+      assert(_offset + _base._headerSize <= buffer.count)
+      let p = buffer.baseAddress! + _offset
+      let header = p.assumingMemoryBound(to: CInterop.CMsgHdr.self)
+      return header.pointee
+    }
+  }
+
+  /// The protocol of the message.
+  public var level: SocketDescriptor.Option.Level {
+    .init(rawValue: _header.cmsg_level)
+  }
+
+  /// The protocol-specific type of the message.
+  public var type: SocketDescriptor.Option {
+    .init(rawValue: _header.cmsg_type)
+  }
+
+  /// Calls `body` with an unsafe raw buffer pointer containing the
+  /// message payload.
+  ///
+  /// This roughly corresponds to the C macro `CMSG_DATA`.
+  ///
+  /// - Note: The buffer passed to `body` does not include storage reserved
+  ///    for holding the message header, such as the `level` and `type` values.
+  ///    To access header information, you have to use the corresponding
+  ///    properties.
+  public func withUnsafeBytes<R>(
+    _ body: (UnsafeRawBufferPointer) throws -> R
+  ) rethrows -> R {
+    try _base._withUnsafeBytes { buffer in
+      let headerSize = _base._headerSize
+      assert(_offset + headerSize <= buffer.count)
+      let p = buffer.baseAddress! + _offset
+      let header = p.assumingMemoryBound(to: CInterop.CMsgHdr.self)
+      let data = p + headerSize
+      let count = Swift.min(Int(header.pointee.cmsg_len) - headerSize,
+                            buffer.count)
+      return try body(UnsafeRawBufferPointer(start: data, count: count))
+    }
+  }
+}
+
+// @available(macOS 9999, iOS 9999, watchOS 9999, tvOS 9999, *)
+extension SocketDescriptor {
+
+  public func sendMessage(
+    bytes: UnsafeRawBufferPointer,
+    recipient: SocketAddress? = nil,
+    ancillary: AncillaryMessageBuffer? = nil,
+    flags: MessageFlags = [],
+    retryOnInterrupt: Bool = true
+  ) throws -> Int {
```

Comment:
Sink this body into an internal function, make it `@usableFromInline`, and always emit this function. Our ABI is `Result`, our API is `throws`.